core/imt/inc/ROOT/TExecutor.hxx

```diff
@@ -0,0 +1,468 @@
+   //
+   template<class F, class Cond = noReferenceCond<F>>
+   auto MapImpl(F func, unsigned nTimes) -> std::vector<typename std::result_of<F()>::type>;
+   template<class F, class INTEGER, class Cond = noReferenceCond<F, INTEGER>>
+   auto MapImpl(F func, ROOT::TSeq<INTEGER> args) -> std::vector<typename std::result_of<F(INTEGER)>::type>;
+   template<class F, class T, class Cond = noReferenceCond<F, T>>
+   auto MapImpl(F func, std::vector<T> &args) -> std::vector<typename std::result_of<F(T)>::type>;
+   template<class F, class T, class Cond = noReferenceCond<F, T>>
+   auto MapImpl(F func, const std::vector<T> &args) -> std::vector<typename std::result_of<F(T)>::type>;
+
+   // Extension of the Map interfaces with chunking, specific to this class and
+   // only available from a MapReduce call.
+   template<class F, class R, class Cond = noReferenceCond<F>>
+   auto Map(F func, unsigned nTimes, R redfunc, unsigned nChunks) -> std::vector<typename std::result_of<F()>::type>;
+   template<class F, class INTEGER, class R, class Cond = noReferenceCond<F, INTEGER>>
+   auto Map(F func, ROOT::TSeq<INTEGER> args, R redfunc, unsigned nChunks) -> std::vector<typename std::result_of<F(INTEGER)>::type>;
+   template<class F, class T, class R, class Cond = noReferenceCond<F, T>>
+   auto Map(F func, std::vector<T> &args, R redfunc, unsigned nChunks) -> std::vector<typename std::result_of<F(T)>::type>;
+   template<class F, class T, class R, class Cond = noReferenceCond<F, T>>
+   auto Map(F func, const std::vector<T> &args, R redfunc, unsigned nChunks) -> std::vector<typename std::result_of<F(T)>::type>;
+   template<class F, class T, class R, class Cond = noReferenceCond<F, T>>
+   auto Map(F func, std::initializer_list<T> args, R redfunc, unsigned nChunks) -> std::vector<typename std::result_of<F(T)>::type>;
+
+private:
+   ROOT::Internal::ExecutionPolicy fExecPolicy;
+#ifdef R__USE_IMT
+   std::unique_ptr<ROOT::TThreadExecutor> fThreadExecutor;
+#else
+   #define fThreadExecutor fSequentialExecutor
+#endif
+#ifndef R__WIN32
+   std::unique_ptr<ROOT::TProcessExecutor> fProcessExecutor;
+#else
+   #define fProcessExecutor fSequentialExecutor
+#endif
+   std::unique_ptr<ROOT::TSequentialExecutor> fSequentialExecutor;
+
+   /// \brief Helper class to get the correct return type from the Map function,
+   /// necessary to infer the ResolveExecutorAndMap function type
+   template<class F, class T>
+   struct MapRetType {
+      using type = typename std::result_of<F(typename T::value_type)>::type;
+   };
+
+   template<class F>
+   struct MapRetType<F, unsigned> {
+      using type = typename std::result_of<F()>::type;
+   };
+
+
+   /// \brief Function called from Map to select and execute the correct Executor
+   /// according to the set Execution Policy.
+   template<class F, class T>
+   auto ResolveExecutorAndMap(F func, T args) -> std::vector<typename MapRetType<F, T>::type> {
```

Comment:
I can't do const, because then I would break the case where the user provides a function that modifies the input parameters, which has been supported until now.

Reply:
`T&&` + `std::forward` then?

Reply:
Sure. I had to decay T in the trailing return type to pass it to MapRetType